packages/grpc-js/src/server.ts

```diff
@@ -1421,41 +1420,72 @@ export class Server {
+          const keepaliveTimeout = setTimeout(() => {
+            if (keepaliveInterval) {
+              clearInterval(keepaliveInterval);
+              keepaliveInterval = null;
+              sessionClosedByServer = true;
+              this.trace('Connection dropped by keepalive timeout');
+              session.close();
+            }
           }, this.keepaliveTimeoutMs);
-          keepaliveTimeoutTimer.unref?.();
+          keepaliveTimeout.unref?.();
 
           try {
-            session.ping(
-              (err: Error | null, duration: number, payload: Buffer) => {
-                if (keepaliveTimeoutTimer) {
-                  clearTimeout(keepaliveTimeoutTimer);
+            if (
+              !session.ping(
+                (err: Error | null, duration: number, payload: Buffer) => {
+                  clearTimeout(keepaliveTimeout);
+                  if (err) {
+                    if (keepaliveInterval) {
+                      clearInterval(keepaliveInterval);
+                      keepaliveInterval = null;
+                    }
+                    sessionClosedByServer = true;
+                    this.trace(
+                      'Connection dropped due to error with ping frame ' +
+                        err.message +
+                        ' return in ' +
+                        duration
+                    );
+                    session.close();
+                  }
                 }
-
-                if (err) {
-                  sessionClosedByServer = true;
-                  this.trace(
-                    'Connection dropped due to error of a ping frame ' +
-                      err.message +
-                      ' return in ' +
-                      duration
-                  );
-                  session.close();
-                }
-              }
-            );
+              )
+            ) {
+              throw new Error('Server keepalive ping send failed');
+            }
           } catch (e) {
-            clearTimeout(keepaliveTimeoutTimer);
-            // The ping can't be sent because the session is already closed
+            // The ping can't be sent because the session is already closed, max outstanding pings reached, etc
+            clearTimeout(keepaliveTimeout);
+            if (keepaliveInterval) {
+              clearInterval(keepaliveInterval);
+              keepaliveInterval = null;
+            }
+            this.trace(
+              'Connection dropped due to error sending ping frame ' +
+                (e instanceof Error ? e.message : 'unknown error')
+            );
             session.destroy();
           }
         }, this.keepaliveTimeMs);
-        keeapliveTimeTimer.unref?.();
+        keepaliveInterval.unref?.();
       }
 
+      session.once('goaway', (errorCode, lastStreamID, opaqueData) => {
+        if (errorCode === http2.constants.NGHTTP2_ENHANCE_YOUR_CALM) {
+          this.trace('Connection dropped by client due to ENHANCE_YOUR_CALM');
```

Comment:
Does the client ever actually send a GOAWAY with this error code? There's a reason for the server to send it: the client can take that information into consideration after reconnecting, and send pings less frequently (and we actually do that). But the server doesn't keep track of which clients have sent this, so it doesn't change behavior.

Also, ENHANCE_YOUR_CALM only means that the connection closed due to excessive pings if the `opaqueData` is equal to the string `too_many_pings`.

In addition, the log lines in both branches are essentially the same. I'd rather be consistent, and either always report the error code, or always report the string name of the error code, and avoid the branch entirely.